misc/releaser.py

```diff
@@ -30,27 +32,126 @@ class ReleaseError(Exception):
+
+        def _pre_type_to_val(pre_type):
+            return {"a": 1, "b": 2, "rc": 3}[pre_type]
+
+        if k == other_k:
+            # Must take into account dev and pre info
+            if self.is_preversion:
+                pre_type = _pre_type_to_val(self.pre_type)
+                pre_index = self.pre_index
+            else:
+                pre_type = pre_index = math.inf
 
+            if other.is_preversion:
+                other_pre_type = _pre_type_to_val(other.pre_type)
+                other_pre_index = other.pre_index
+            else:
+                other_pre_type = other_pre_index = math.inf
 
-def parse_version(raw):
-    raw = raw.strip()
-    if not raw.startswith("v"):
-        raw = f"v{raw}"
-    match = re.match(r"^v([0-9]+)\.([0-9]+)\.([0-9]+)(([\-+]\w+)*)$", raw)
-    if match:
-        major, minor, patch, extra, *_ = match.groups()
-        return Version(raw, f"{major}.{minor}.{patch}", int(major), int(minor), int(patch), extra)
-    else:
-        raise ValueError(
-            "Invalid version format, should be `[v]<major>.<minor>.<patch>[-<extra>]` (e.g. `v1.0.0`, `1.2.3-dev`)"
-        )
+            dev = 1 if self.is_dev else 0
+            other_dev = 1 if other.is_dev else 0
+
+            return (pre_type, pre_index, dev) < (other_pre_type, other_pre_index, other_dev)
+
+        else:
+            return k < other_k
+
+    def __le__(self, other):
+        if self == other:
+            return True
+        else:
+            return self < other
+
+
+# Version is non-trivial code, so let's run some pseudo unit tests...
+def _test(raw, is_dev=False, pre=(None, None)):
+    v = Version(raw)
+    assert v.is_dev is is_dev
+    pre_type, pre_index = pre
+    assert v.is_preversion is (pre_type is not None)
+    assert v.pre_type == pre_type
+    assert v.pre_index == pre_index
+
+
+_test("1.2.3")
+_test("1.2.3+dev", is_dev=True)
+_test("1.2.3-10-g3b5f5762", is_dev=True)
+_test("1.2.3-b42", pre=("b", 42))
+_test("1.2.3-rc1+dev", is_dev=True, pre=("rc", 1))
+assert Version("1.2.3") < Version("1.2.4")
+assert Version("1.2.3-a1") < Version("1.2.3-a2")
+assert Version("1.2.3-a10") < Version("1.2.3-b1")
+assert Version("1.2.3-b10") < Version("1.2.3-rc1")
+assert Version("1.2.3-b1") < Version("1.2.3-b1+dev")
+assert Version("1.2.3-rc1") < Version("1.2.3")
+assert Version("1.2.3") < Version("1.2.3+dev")
+assert Version("1.2.3+dev") < Version("1.2.4-rc1")
+assert Version("1.2.4-rc1") < Version("1.2.4-rc1+dev")
+assert Version("1.2.4-rc1+dev") < Version("1.2.4")
+assert Version("1.2.3") < Version("1.2.3-10-g3b5f5762")
+assert Version("1.2.3-10-g3b5f5762") < Version("1.2.4-b42")
+assert Version("1.2.4-b42") < Version("1.2.4-b42-10-g3b5f5762")
+assert Version("1.2.4-b42-10-g3b5f5762") < Version("1.2.4")
+assert Version("1.2.3-rc10+dev") < Version("1.2.3")
+assert Version("1.2.3-b10+dev") < Version("1.2.3-rc1+dev")
+assert Version("1.2.3-b10+dev") < Version("1.2.3+dev")
```

Comment:
This should go to a `test_version` function so we can run `pytest misc/releaser.py`

Reply:
The thing is releaser is use in the Q&A job of the CI, so in parallel with the Unittest job...

I think it's better to keep the asserts inline in the file given there is not much of them (so doesn't clutter too much the file and doesn't slow down the speed) and the script is not supposed to be modified often. This way we have a standalone script that ensure it integrity as part of it run ;-)